xarray/core/dataset.py

```diff
@@ -4149,6 +4163,108 @@ def reset_index(
 
         return self._replace(variables, coord_names=coord_names, indexes=indexes)
 
+    def set_xindex(
+        self,
+        coord_names: Hashable | Sequence[Hashable],
+        index_cls: type[Index] | None = None,
+        **options,
+    ) -> Dataset:
```

Comment:
```suggestion
    ) -> T_Dataset:
```

Reply:
mypy is not happy with this:

```
xarray/tests/test_dataset.py:3307: error: Argument 1 to "set_xindex" of "Dataset" has incompatible type "List[str]"; expected "Hashable"  [arg-type]
xarray/tests/test_dataset.py:3307: note: Following member(s) of "List[str]" have conflicts:
xarray/tests/test_dataset.py:3307: note:     __hash__: expected "Callable[[], int]", got "None"
xarray/tests/test_dataset.py:3307: note: Protocol member Hashable.__hash__ expected instance variable, got classe variabl
```

https://github.com/pydata/xarray/pull/6971#discussion_r959746215

Reply:
strings are sequences apparently:
```python
isinstance("str", typing.Sequence)
Out[63]: True
```
Try out `CoordNames = Union[str, Iterable[Hashable]]` seems to be succesful in #7048.
It would be nice if we aligned these tricky types so try to use named variables for repeated arguments.

Reply:
Technically a str is also an Iterable of Hashable :P
But the typing community is quite relaxed about violating that fact.
So as long as you don't need the two types to be "perpendicular" it should work.

Reply:
+1 for using a named variable like `CoordNames`. The tricky thing here is that the order is important. Do we use `Sequence` in Xarray in that case? I guess we would need to define two variables for each case where the order does / doesn't matter?

Also, I don't remember whether a single coordinate name should be `str` or `Hashable`. Should we treat it like a single dimension name or not?

I feel like this issue should be addressed more globally in Xarray than within the scope of this PR. Perhaps better to move on and merge this PR before the next release?

Reply:
Usually we try to move to `str | Iterable [Hashable]` for "one or more dims", and `Hashable` for a single dim.

Reply:
> Usually we try to move to str | Iterable[Hashable] for "one or more dims", and Hashable for a single dim.

Probably not in all cases? For example, with `DataArray.__init__(..., dims: str | Iterable[Hashable])` the type checker would allow passing a set. Recently I had to figure out what was going on with `xr.DataArray(data=np.zeros((10, 5)), dims={'x', 'time'})`, which mypy should actually catch with `Sequence[Hashable]`. Slightly off-topic: should we have two variables `Dims` and `OrderedDims` defined in `xarray.core.types`?

Same issue here for coordinate names. `str | Sequence[Hashable]` seems to work well, though.

Reply:
Why should a set not be allowed?
It's already since quite some time that the order is preserved? I think all built-in Iterables have conserved order, and internally we convert to tuple anyway?

Reply:
I don't think the order is preserved for sets (unlike dicts). This is what I can get with CPython 3.9 / Xarray v2022.6.0:

```python
print(xr.DataArray(data=np.zeros((2, 3)), dims={'x', 'time'}))
# <xarray.DataArray (time: 2, x: 3)>
# array([[0., 0., 0.],
#        [0., 0., 0.]])
# Dimensions without coordinates: time, x

tuple({'x', 'time'})
# ('time', 'x')
```

Reply:
Whoops you are right, that was dicts.
Then indeed we need to distinguish between dims and ordered dims.